spec/consensus/light-client.md

```diff
@@ -1,318 +1,577 @@
+    // this ensures that there is at least single correct validator in the set of signers
+    if signedPower < max(1/3, trustLevel) * totalPower return ErrInsufficientVotingPower
+    return nil
+}
 
-This can be used in several settings:
+// returns nil if commit is signed by more than 2/3 of voting power of the given validator set
+// return error otherwise
+func verifyCommitFull(vs ValidatorSet, commit Commit) error {
+    totalPower := vs.TotalVotingPower;
+    signed_power := votingPowerIn(signers(commit, vs), vs)
 
-- someone tells the lite client that application data that is relevant for it can be read in the block of height _k_.
-- the lite clients wants the latest state. It asks a full node for the current height, and uses the response for _k_.
+    // check the signers account for +2/3 of the voting power
+    if signed_power * 3 <= total_power * 2 return ErrInvalidCommit
+    return nil
+}
+```
 
-## Details
+**VerifyHeaderAtHeight.** The function `VerifyHeaderAtHeight` captures high level
+logic, i.e., application call to the lite client module to download and verify header
+for some height.
 
-_Assumptions_
+```go
+func VerifyHeaderAtHeight(untrustedHeight int64,
+                          trustThreshold TrustThreshold,
+                          trustingPeriod Duration,
+                          clockDrift Duration,
+                          trustedState TrustedState) (error, TrustedState)) {
 
-1. _tp < unbonding period_.
-2. _snh.Header.bfttime < now_
-3. _snh.Header.bfttime < h.Header.bfttime+tp_
-4. _trust(h)=true_
+    trustedHeader := trustedState.SignedHeader.Header
 
-**Observation 1.** If _h.Header.bfttime + tp > now_, we trust the old
-validator set _h.Header.NextV_.
+    now := System.Time()
+    if !isWithinTrustedPeriod(trustedHeader, trustingPeriod, now) {
+        return (ErrHeaderNotWithinTrustedPeriod, trustedState)
+    }
 
-When we say we trust _h.Header.NextV_ we do _not_ trust that each individual validator in _h.Header.NextV_ is correct, but we only trust the fact that at most 1/3 of them are faulty (more precisely, the faulty ones have at most 1/3 of the total voting power).
+    newTrustedState, err := VerifyAndUpdateBisection(trustedState,
+                                                     untrustedHeight,
+                                                     trustThreshold,
+                                                     trustingPeriod,
+                                                     clockDrift,
+                                                     now)
 
-### Functions
+    if err != nil return (err, trustedState)
 
-The function _Bisection_ checks whether to trust header _h2_ based on the trusted header _h1_. It does so by calling
-the function _CheckSupport_ in the process of
-bisection/recursion. _CheckSupport_ implements the trusted period method and, for two adjacent headers (in term of heights), it checks uninterrupted sequence of proof.
+    now = System.Time()
```

Comment:
Do we really want the function to have access to the system clock ?

Reply:
The idea of this function is to illustrate how to correctly use `VerifyBisection` function. In case we don't check if we are still within trusted period of initial trusted state after `VerifyBisection` is executed, we can't give precise guarantees to the user. I don't know how to avoid this. As most of complexity is happening within `VerifyBisection` this shouldn't be a big problem from the testing perspective.